### ISSUE TITLE
Create a formatter based on Angular components

### ISSUE THREAD

**Issue by sabeurch** [ghiscoding/Angular-Slickgrid#7]
Looking into the source code that shows how to create formatters, I found out how to create a PercentCompleteFormatter formatter, I brought it into my project and it works fine.

> `export const PercentCompleteFormatter: Formatter = (row: number, cell: number, value: any, columnDef: Column, dataContext: any): any => {
>     // OK
>     if ( value === null || value === '' ) {
>         return '-';
>     } else if ( value < 50 ) {
>         return "<span style='color:red;font-weight:bold;'>${value}%</span>";
>     } else {
>         return "<span style='color:green'>${value}%</span>";
>     }
> }`

Would you please tell me how to use my custom Angular components instead of using HTML tag "<span>" ?
If not possible, is there any Angular like approach ?
Thanks,

**Comment by ghiscoding** [comment]
Most of the functionalities have a Wiki explaining how to use them, I also have Wiki links in each demo page. This is the [Formatters Wiki](https://github.com/ghiscoding/Angular-Slickgrid/wiki/Formatters) and all the info is in there. 

As to `can I use something else than plain html/javascript?` The answer would unfortunately be NO, at least not at the moment, this library is really a wrapper of the original jQuery Slickgrid and I don't intend to change the original implementation of a SlickGrid Formatter. The problem also is that SlickGrid expect the Formatter to be synchronous and if you don't return a result right away then you'll have problem with displaying data to the cell value. However SlickGrid also provide an `asyncPostRender` for asynchronous rendering of a Formatter, [here](http://6pac.github.io/SlickGrid/examples/example10a-async-post-render-cleanup.html) is an example, if you look at the graph, it is populated afterward. 

Anyhow, I'm not completely close to the idea if you want to spend some time and create a PR for it. However I won't spend time myself on this, since I have other priorities and other features in the queue. If you want to take a look at it, you will most probably have to use `asyncPostRender` since you'll have to render the Angular template first, then display it in SlickGrid.

**Comment by sabeurch** [comment]
Thanks Ghislain for the good job, 
For the moment I am going to get limited into using HTML components  (ready rendered).
One other thing, is it possible to include further 6pac functionalities into this "Angular-Slickgrid" (in case where they are not, example "rows grouping"), any developer documentations are available ?

**Comment by ghiscoding** [comment]
I'm also curious, what kind of Angular Template that you want to do which you can't do in a regular html/js Formatter?

**Comment by sabeurch** [comment]
Dear  Ghislain, 
I am convinced with the use of html/js formatter for the moment.
My last question was about: Should I consider that the library "Angular-Slickgrid" contains all functionalities inside the "6pac/SlickGrid" ?
For example: Select box editor
http://6pac.github.io/SlickGrid/examples/example-select2-editor.html
and also rows grouping (Advanced datagrid style): http://6pac.github.io/SlickGrid/examples/example5-collapsing.html
How to add them into this library "ghiscoding/Angular-Slickgrid" ? or to use them since as you said "this library is really a wrapper of the original jQuery Slickgrid " ?

**Comment by ghiscoding** [comment]
Yes the plan is to eventually have all features of 6pac fork included in Angular-Slickgrid, but it all goes by priorities and when I need them at work. This project is mostly done during working hours, we decided to make it open source to get feedback and hopefully some help on implementing certain features (like Angular template would be nice). 

The next features that are in my queue, before even grouping, are:
1- export to CSV or text file
2- reload a grid with certain preset (like filters and possibly sorts)

The number 2 is probably a big one, so I'm not sure when grouping will come in play, surely not on the short term.
When you say yourself, did you mean that you want to take a look at possibly implementing some of the features? If so, you can give it a shot and I can guide you. I try to create separate services for each features which is easier to work with and refer to, you can see the list of current [services](https://github.com/ghiscoding/Angular-Slickgrid/tree/master/src/app/modules/angular-slickgrid/services), Grouping will have to be a separate service as well, so you can start with that if you want.

**Comment by ghiscoding** [comment]
Thinking about it, you might be able to implement some of the SlickGrid features directly on the Grid Object (or DataView object). I provide these objects through an EventEmitter and you could technically call any SlickGrid stuff you want on these objects. Take a look at these emitter on the [Wiki - Grid & DataView Events](https://github.com/ghiscoding/Angular-Slickgrid/wiki/Grid-&-DataView-Events).

For example
```html
<angular-slickgrid 
  gridId="grid2" 
  (dataviewChanged)="dataviewReady($event)" 
  (gridChanged)="gridReady($event)"
  [columnDefinitions]="columnDefinitions" 
  [gridOptions]="gridOptions" 
  [dataset]="dataset">
</angular-slickgrid>
```
```javascript
  gridReady(grid) {
    // do stuff with the grid object
  }
  dataviewReady(dataview) {
    // do stuff with the dataView object
  }
You then have access to the SlickGrid original objects, I provided them so that you can hook to them and do what you want the old SlickGrid way. So you can do that until I wrap the feature in the future.

**Comment by sabeurch** [comment]
Cool and perfect, thank you !

**Comment by ghiscoding** [comment]
Please note that I just released a new version [0.7.10](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v0.7.10) and I decided to renamed the 2 Event Emitters that I told you about in my last comment. You can see the [Wiki - Grid & DataView Events](https://github.com/ghiscoding/Angular-Slickgrid/wiki/Grid-&-DataView-Events) documentation.

and here's a portion of the new release that affects you
- added many new Event Emitters
  - `onDataviewCreated`
  - `onGridCreated`
  - `onBeforeGridCreate`
  - `onBeforeGridDestroy`
  - `onGridDestroyed`

#### OBSOLETE in future version
These 2 Event Emitters got renamed to be consistent in the new naming structure (every emitter now start with onX prefix). You can still use the old name, but I'd like to remove then in future releases.
- `dataviewChanged` renamed to `onDataviewCreated`
- `gridChanged` renamed to `onGridCreated`

**Comment by sabeurch** [comment]
Thank you  Ghislain, It is interesting to know, I'll have to update the library.
I guess if there is a possibility

**Comment by sabeurch** [comment]
I really appreciate this great job.
Upgraded to version 0.7.10 and it works fine.
One important feature that looks to be helpful: the "params" attribute in a column definition.
When defining my column, I can set a background color function, it return a color name based on a business being implemented in my interface (entries are field and dataContext)
```
 params: {
        backgroundColorFunction: this.cellColorFunction
}
```
This will be retrieved from the formatter body as follow:
```
export const DurationFormatter: Formatter = (row: number, cell: number, value: any, columnDef: Column, dataContext: any): any => {
    let color = columnDef.params.backgroundColorFunction(dataContext, columnDef.field);
    if ( value === null || value === '' ) {
        return '-';
    } else if ( value > 50 ) {
        return `<div style='color:yellow;font-weight:bold;background-color:${color};width:100%;height:100%'>${value}%</div>`;
    } else {
        return `<span style='color:black'>${value}%</span>`;
    }
};
```
I guess, this may allow me later to inject any DOM objects.

**Comment by ghiscoding** [comment]
Yes the `params` in Column Definition is a `any` type and was created to do exactly that, pass extra arguments to your Formatters. I created it when I wanted to pass a Translate Service instance to translateFormatter that I created. So your assumptions are correct.

To go back on the original ticket, it would be kinda hard to make an Angular Component work with a SlickGrid Formatter. We would have to first compile the component into plain html/js code in order to use it in a SlickGrid Formatter, these Formatter are strictly returning a value string or html markup that SlickGrid can then evaluate. 

I'm still new to Angular 4 (we were still using AngularJS at work couple months ago) and I haven't seen, neither searched for, ways to compile component to html/js. I also wonder what this feature would bring to the table. What type of component did you have in mind, do you have any examples?

Side note, if you have question and/or issues that are different from the original issue, it might be better to create new issue(s). It's easier to read and handle :)  There's a lot of WiKi but if the info is not clear enough or is not there, feel free to ask and/or do PR (even for WiKi). Thanks

**Comment by ghiscoding** [comment]
@sabeurch I'm just trying to clear up some of the things you wrote in this long thread. 

> My last question was about: Should I consider that the library "Angular-Slickgrid" contains all functionalities inside the "6pac/SlickGrid" ?
For example: Select box editor http://6pac.github.io/SlickGrid/examples/example-select2-editor.html
and also rows grouping (Advanced datagrid style): http://6pac.github.io/SlickGrid/examples/example5-collapsing.html

When you wrote about `Select box editor` in `example-select2-editor.html`, did you mean Custom Inline Editor?
- If so, I added more documentation on how to create [Wiki - Custom Inline Editors](https://github.com/ghiscoding/Angular-Slickgrid/wiki/Editors)
- I released a new version [0.10.0](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v0.10.0) today which also brings [MultipleSelect Filter](https://github.com/ghiscoding/Angular-Slickgrid/wiki/Select-Filter#filtertypemultipleselect) and even [Custom Filter](https://github.com/ghiscoding/Angular-Slickgrid/wiki/Custom-Filter). 

Grouping is not planned at all in the near feature but you can do it directly with the SlickGrid object as mentioned earlier.

**Comment by ghiscoding** [comment]
The thread is getting quite long but...

For the original question, if anyone wants to look into this. This is probably doable through [Angular Universal](https://universal.angular.io/) which is SSR (Server Side Rendering). With that lib, we can apparently create the DOM string which we could then pass to a SlickGrid Formatter.

This is very very low priority for me, if anyone wants to take a look at it, I certainly welcome PR.

**Comment by sabeurch** [comment]
Indeed, it is now a non urgent feature, not really needed for me anymore.
I think it's Ok to close this thread.

**Comment by ghiscoding** [comment]
Spring cleaning, I probably will never put this in but if that ever happen I will re-open this ticket. Thanks for your feedback

**Comment by ghiscoding** [comment]
Reopening as this will be possible in next version, however there are limitation. This is doable via `postRender` but this is an async function in SlickGrid and this works fine but we can visually see the data being change. So basically, we cannot use a Formatter because Angular Component needs at least 1 cycle before the Angular Component is rendered properly and SlickGrid Formatter requires string right away, so we cannot use Formatter for that reason. 

Here's a [SlickGrid PostRender demo](http://6pac.github.io/SlickGrid/examples/example10-async-post-render.html) and here's an animated gif of an Angular Component using the `postRender`. While I scroll up down, you can visually see a slight delay in rendering the content of that cell (the bold text, 3rd column)... but hey it's works and it might useful for some. 

![2019-02-19_21-07-59](https://user-images.githubusercontent.com/643976/53061045-87346580-348a-11e9-95f3-dfe33be6e966.gif)

**Comment by ghiscoding** [comment]
Now available under version [2.3.0](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v2.3.0)
